[AIRFLOW-3231] Basic operators for Google Cloud SQL

sprzedwojski opened the pull request:
Add CloudSqlInstanceDeployOperator, CloudSqlInstancePatchOperator and CloudSqlInstanceDeleteOperator.

Each operator includes:
- core logic
- input params validation
- unit tests
- presence in the example DAG
- docstrings
- How-to and Integration documentation

Additionally, small improvements to GcpBodyFieldValidator were made:
- add simple list validation capability (type="list")
- introduced parameter allow_empty, which can be set to False
	to test for non-emptiness of a string instead of specifying
	a regexp.

TrevorEdwards commented on airflow/contrib/operators/gcp_sql_operator.py:
```diff
@@ -0,0 +1,299 @@
+        dict(name="tier", allow_empty=False),
+        dict(name="backupConfiguration", type="dict", fields=[
+            dict(name="binaryLogEnabled", optional=True),
+            dict(name="enabled", optional=True),
```
Have you considered using required=True (but defaults to False) instead of optional=True?

I think this discussion can easily go either way, so I'm fine with using optional=True. I think required=True might be more useful if we assume the most fields are going to be optional (smaller validator specs), and most users want to set the minimum amount of fields (e.g. ctrl + f for required=True)

sprzedwojski replied:
I talked to Jarek and in other cases there can be more fields that are required than optional. We would prefer to leave it as is for now, so that we don't have to update the already merged operators using the validator.
We could revisit this in the future and potentially refactor all the operators in one PR.

TrevorEdwards commented on airflow/contrib/operators/gcp_sql_operator.py:
```diff
@@ -0,0 +1,299 @@
+    use CloudSqlInstanceDeployOperator.
+
+    In the request body, supply the relevant portions of an instance resource, according
+    to the rules of patch semantics.
```
I don't think there's a single rule of patch semantics (I know this came from CloudSQL docs, but I find that sentence too vague).

Let's link to https://cloud.google.com/sql/docs/mysql/admin-api/how-tos/performance#patch right after this sentence.

sprzedwojski replied:
You're right, taken out of the context of the Cloud SQL docs this sentence sounded a bit weird here.
I've added the link.

sprzedwojski commented on docs/howto/operator.rst:
```diff
@@ -283,3 +283,170 @@ See `Adding the IAM service agent user role to the runtime service <https://clou
+
+.. literalinclude:: ../../airflow/contrib/example_dags/example_gcp_sql.py
+    :language: python
+    :start-after: [START howto_operator_cloudsql_arguments]
```
In the source code this snippet is already maximally to the left, so dedenting it would cut into the text.
```
# [START howto_operator_cloudsql_arguments]
PROJECT_ID = models.Variable.get('PROJECT_ID', '')
INSTANCE_NAME = models.Variable.get('INSTANCE_NAME', '')
# [END howto_operator_cloudsql_arguments]
```

sprzedwojski commented on docs/howto/operator.rst:
```diff
@@ -283,3 +283,170 @@ See `Adding the IAM service agent user role to the runtime service <https://clou
+
+.. literalinclude:: ../../airflow/contrib/example_dags/example_gcp_sql.py
+    :language: python
+    :start-after: [START howto_operator_cloudsql_arguments]
```
Same as https://github.com/PolideaInternal/incubator-airflow/pull/8#discussion_r227433302

sprzedwojski commented on docs/integration.rst:
```diff
@@ -501,6 +501,37 @@ BigQueryHook
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autoclass:: airflow.contrib.operators.gcp_sql_operator.CloudSqlInstancePatchOperator
+
```
Done

sprzedwojski commented on airflow/contrib/operators/gcp_sql_operator.py:
```diff
@@ -0,0 +1,299 @@
+    """
+    # [START gcp_sql_deploy_template_fields]
+    template_fields = ('project_id', 'name', 'gcp_conn_id', 'api_version',
+                       'validate_body')
```
Removed

sprzedwojski commented on airflow/contrib/hooks/gcp_sql_hook.py:
```diff
@@ -0,0 +1,191 @@
+
+        :param project_id: Project ID of the project where the instance will be created.
+        :type project_id: str
+        :param body: Body required by the Cloud SQL insert API, as described in TODO
```
Done

sprzedwojski commented on airflow/contrib/hooks/gcp_sql_hook.py:
```diff
@@ -0,0 +1,191 @@
+
+        :param project_id: Project ID of the project where the instance exists.
+        :type project_id: str
+        :param body: Body required by the Cloud SQL update API, as described in TODO
```
@potiuk I've replaced the TODO (sorry!) with a URL to the docs
@TrevorEdwards I've added a `Caution` notice about overriding the settings, similar to the one in the operator's docstring (both based on the Cloud SQL documentation)

sprzedwojski commented on airflow/contrib/example_dags/example_gcp_sql.py:
```diff
@@ -0,0 +1,137 @@
+# [END howto_operator_cloudsql_arguments]
+
+# [START howto_operator_cloudsql_deploy_body]
+body = {
```
Done

sprzedwojski commented on airflow/contrib/example_dags/example_gcp_sql.py:
```diff
@@ -0,0 +1,137 @@
+    "region": "europe-west4",
+}
+# [END howto_operator_cloudsql_deploy_body]
+update_body = {
```
In theory it should, however I found some inconsistencies (e.g. storage size and type is not reset to default, but other settings such as `maintenaceWindow` is) - I mentioned this in an email to Aastha yesterday.

sprzedwojski commented on airflow/contrib/example_dags/example_gcp_sql.py:
```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
```
I will change it when rebasing for the final PR.

sprzedwojski commented:
I've created a separate [PR](https://github.com/apache/incubator-airflow/pull/4092) fixing the issue with flake8 tests failing on Travis. It already got merged and I rebased this PR to include the fix.

Apart from that I also noticed there were inconsistencies in variable names between our operators and the Python Google API / docs (namely we had `name` for the instance name, when the docs have `instance`, similarly we had `project_id` while in the docs it was simply `project`).

That's why I also added a fixup to unify all the names. These changes are minor, but feel free to take a look if you wish before I close this PR and create the one to the `apache/incubator-airflow` repo.

TrevorEdwards commented:
I'm leaning toward sticking with project_id despite the API using "project".

1. It's consistent with most GCP operators (there have previously been inconsistency issues)
2. The docs suggest it only accepts project_id as opposed to project_number, so calling it project_id should reduce cognitive load.

sprzedwojski commented:
Closing this internal PR, will squash and open a new one to `apache`.
